JavaScript/avl.js

```diff
@@ -0,0 +1,317 @@
+            // Node has right child present or it does not have both childs
+            if (node.left == null) {
+                let temp = node.right;
+                node = null;
+                return temp;
+            }
+
+            // Node has left child present
+            else if (node.right == null) {
+                let temp = node.left;
+                node = null;
+                return temp;
+            }
+
+            // Node has both child present
+            // Get the in-order successor element in the right subtree, and replace
+            // it by the node to be deleted and delete the in-order successor.
+            let temp = this.getInorderSuccessor(node.right);
+            node.value = temp.value;
+            node.right = this.deleteHelper(node.right, temp.value);
+        }
+
+        // Update the heights of the node after deletion
+        node.height = 1 + Math.max(this.getHeight(node.left), this.getHeight(node.right));
+
+        // Rebalance the node
+        return this.updateBalance(node, node.value);
+    }
+
+    /**
+     * Gets the In-Order successor of the given node
+     * (node with the minimum value in the left subtree.)
+     *
+     * @param node Node whose left subtree is to be searched.
+     * @return in-order successor.
+     */
+    getInorderSuccessor(node) {
+        if(node == null || node.left == null)
+            return node;
+        return this.getInorderSuccessor(node.left);
+    }
+
+    /**
+     * Prints the tree using In-Order traversal.
+     * @return
+     */
+    inOrder() {
+        let inOrderList = [];
+        this.inOrderHelper(this.root, inOrderList);
+        return inOrderList;
+    }
+
+    inOrderHelper(node, inOrderList) {
+        if(node != null){
+            this.inOrderHelper(node.left, inOrderList);
+            inOrderList.push(node.value);
+            this.inOrderHelper(node.right, inOrderList);
+        }
+    }
+
+    /**
+     * Prints the tree using Post-Order traversal.
+     * @return
+     */
+    postOrder() {
+        let postOrderList = [];
+        this.postOrderHelper(this.root, postOrderList);
+        return postOrderList;
+    }
+
+    postOrderHelper(node, postOrderList) {
+        if(node != null){
+            this.postOrderHelper(node.left, postOrderList);
+            this.postOrderHelper(node.right, postOrderList);
+            postOrderList.push(node.value);
+        }
+    }
+
+    /**
+     * Prints the tree using Pre-Order traversal.
+     * @return
+     */
+    preOrder() {
+        let preOrderList = [];
+        this.preOrderHelper(this.root, preOrderList);
+        return preOrderList;
+    }
+
+
+    preOrderHelper(node, preOrderList) {
+        if(node != null){
+            preOrderList.push(node.value);
+            this.preOrderHelper(node.left, preOrderList);
+            this.preOrderHelper(node.right, preOrderList);
+        }
+    }
+}
+
+
+// Sample test cases to test the AVL Tree Implementation.
```

Comment:
Next time, bind them in a single function/block. These statements will be executed when the file is imported.

Reply:
Sure thing.